### PR TITLE
adding Meta impl

### PR DESCRIPTION
**Pull request by jondot**
Meta implementation (see https://github.com/SpectralOps/teller/issues/65)

* [x] add types + empty impls for all providers
* [x] add teller providers export command that outputs all providers, and their meta in JSON. Add teller-specific metadata to this JSON (what version of teller generated it, will be useful for docs)
```js
{ 
     "teller":{
         "version": ".."
      },
      "providers":{ 
          "heroku":{
                 ..
          }
      }
}
```
- while exporting, if duplicate provider name found -- warn (this is a way for us to start linting providers)

* [x]  add concrete implementation for every provider for their meta
* [ ] replace config template with the relevant provider meta (this means template is being built real-time and/or different templating generation flow)
* [ ] build a tool to generate README automatically from export JSON, add to workflow as a make command
* [ ] `teller-docs` build a tool to create markdown pages for docs per provider with a standard layout, from exported JSON

### CODE REVIEW

File: pkg/core/types.go [hunk 1]
```diff
@@ -198,6 +199,19 @@ type Provider interface {
...
+	Description    string
+	Authentication string
+	ConfigTemplate string
+	Ops            OpMatrix
+}
+type OpMatrix struct {
+	Sync    bool
+	Mapping bool
+	Read    bool
+	Write   bool
```
Comment (kaplanelad) [inline pkg/core/types.go:212]: We have providers (Consul, etcd for example) that implement only put -> "write" without delete. 
I think we have two options:
1. Split the `Write` to `Put` and `Delete.`
2. `Write` implementation required both of the action types.

WDTY?

Comment (jondot) [reply pkg/core/types.go:212]: Yea, splitting sounds good

**Comment by eladbash** [comment]
https://github.com/tellerops/teller/issues/125
will be implemented on this branch as well

File: pkg/porcelain.go [hunk 2]
```diff
@@ -29,6 +29,9 @@ func (p *Porcelain) StartWizard() (*core.WizardAnswers, error) {
 
 	providers := BuiltinProviders{}
 	providerNames := providers.ProviderHumanToMachine()
+	fmt.Println("Available providers:")
```
Comment (eladbash) [inline pkg/porcelain.go:32]: Remove

File: pkg/providers/vercel.go [hunk 3]
```diff
@@ -59,11 +59,37 @@ func (vp *VercelProject) envMap() map[string]*string {
 https://vercel.com/docs/api#endpoints/secrets
 */
 
-//nolint: golint
-const VERCEL_API_BASE = "https://api.vercel.com/"
+const VercelAPIBase = "https://api.vercel.com/"
+
+const ProjectEndPoint = "/projects"
```
Comment (eladbash) [inline pkg/providers/vercel.go:64]: lint

File: pkg/teller_test.go [hunk 4]
```diff
@@ -68,17 +69,51 @@ func (im *InMemProvider) ProviderHumanToMachine() map[string]string {
...
+		Description: "test-provider",
+	}
+
+	inmemProviderErrorMeta := core.MetaInfo{
+		Name:        "inmem-provider-error",
+		Description: "test-provider-error",
+	}
+
+	providers.RegisterProvider(inmemProviderMeta, NewInMemProvider)
+	providers.RegisterProvider(inmemProviderErrorMeta, NewInMemErrorProvider)
```
Comment (eladbash) [inline pkg/teller_test.go:90]: we used inmem provider only in test, so it's builder wasn't following the builder standard, I changed it so the test will register 2 providers for inmem tests:
1. When we want it to work normally - alwaysError: false
2. When we want it to have error - alwaysError: true
When resolving the first one - we are just using the `inmem-provider` key
When we want to test with errors we resolve the other - `inmem-provider-error` key

**Pull request merged**